python/cpython
# gh-101859: Add caching of `types.GenericAlias` objects

Goals of this PR:
1. Make code using `Py_GenericAlias` C-function cacheable. This is similar to how we use `typing.List[int] is typing.List[int]`
2. Support caching of unpacked generic alias, so `*tuple[int] is *tuple[int]`, but clearly `is not tuple[int]`
3. Do not cache direct `types.GenericAlias` creation. For example, right now `typing._GenericAlias` has this behavior: `typing._GenericAlias(typing.List, (int,)) is not typing._GenericAlias(typing.List, (int,))`. So, I kept the same for `types.GenericAlias`: `types.GenericAlias(list, (int,)) is not types.GenericAlias(list, (int,))`. **Note**: this basically means that all classes using `__class_getitem__ = classmethod(GenericAlias)` are not cached
4. Keeps cache size under control: see `3.`

## Open problems

### References are leaked 🙂 

For example, run: `./python.exe -m test -R 3:3 -v test_genericalias -m test_no_chaining`

```python
test_genericalias leaked [2, 2, 2] references, sum=6
test_genericalias leaked [1, 1, 1] memory blocks, sum=3
test_genericalias failed (reference leak)
```

It will detect a reference leak. As far as I understand the only thing we leak is `genericalias_cache` state object. It is only cleared after the interpreter is shut down.

I see several options here:
1. Add a protected classmethod for `types.GenericAlias` to clear the cache and use it in global test teardown with other cleanup
2. Add a regular module state, for example as `_types` module
3. Find another way to cache it ???
4. Give up on this idea

### Performance and memory testing

I haven't done any yet. But, I think I slowed things down pretty badly :(
But, as far as I understand this is a price we want to pay for less memory usage.

### Final thoughts

I am bit *rusty* (🦀, pun intended) with C code, so any feedback is welcome.


* Issue: gh-101859
